cmd/collector/main.go

```diff
@@ -79,21 +85,33 @@ func main() {
 				logger.Fatal("Failed to create span writer", zap.Error(err))
 			}
 
+			requireLockAndSamplingStore, err := strategyStoreFactory.RequiresLockAndSamplingStore()
+			if err != nil {
+				logger.Fatal("Failed to determine if lock and sampling store is required.", zap.Error(err))
+			}
+			var lock distributedlock.Lock
+			var samplingStore samplingstore.Store
+			if requireLockAndSamplingStore {
+				lock, samplingStore, err = storageFactory.CreateLockAndSamplingStore()
+				if err != nil {
+					logger.Fatal("Failed to create lock and sampling store for adaptive sampling", zap.Error(err))
+				}
+			}
 			strategyStoreFactory.InitFromViper(v)
-			if err := strategyStoreFactory.Initialize(metricsFactory, logger); err != nil {
+			if err := strategyStoreFactory.Initialize(metricsFactory, logger, lock, samplingStore); err != nil {
 				logger.Fatal("Failed to init sampling strategy store factory", zap.Error(err))
 			}
-			strategyStore, err := strategyStoreFactory.CreateStrategyStore()
+			strategyStore, aggregator, err := strategyStoreFactory.CreateStrategyStore()
 			if err != nil {
 				logger.Fatal("Failed to create sampling strategy store", zap.Error(err))
 			}
-
 			c := app.New(&app.CollectorParams{
 				ServiceName:    serviceName,
 				Logger:         logger,
 				MetricsFactory: metricsFactory,
 				SpanWriter:     spanWriter,
 				StrategyStore:  strategyStore,
+				Aggregator:     aggregator,
```

Comment:
Note: we intentionally kept collector unaware of "aggregators", it only needed to know about generic "span processors"

Reply:
I had it this way (and do not mind moving back) but preferred this way for two reasons:

1) collector can own the lifecycle of the Aggregator and call stop when it is closed
2) otherwise the main function has to do the translation from Aggregator -> ProcessSpan which just felt like clutter in main

Reply:
If ProcessSpan was an interface, implementations complying with io.Closer could be closed by the spanProcessor. Without that, I think I also prefer the current code from this PR.